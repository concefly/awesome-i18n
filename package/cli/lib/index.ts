#!/usr/bin/env node
import AI18n, { IConfig } from 'ai18n-core';
import * as diffUtil from 'jsondiffpatch';
import { getUserConfig } from './config';
import * as yargs from 'yargs';

yargs
  .command(
    'check',
    '检查文案变化(可用于 ci)',
    _yargs => {},
    argv => {
      const userConfig = getUserConfig(argv.config as string);

      const stopError = new Error('停止');

      const config: IConfig = {
        ...userConfig,
        hook: {
          afterReduce: (result, extra) => {
            userConfig.hook?.afterReduce?.(result, extra);

            const reduceMsgKeys = [...result.data.keys()].sort();
            const localMsgKeys = Object.keys(extra.localMsgJson).sort();

            // 求 diff
            const diffDelta = diffUtil.diff(localMsgKeys, reduceMsgKeys);
            if (diffDelta) {
              const output = diffUtil.formatters.console.format(diffDelta, undefined);
              throw new Error('文案有变化，请执行 ai18n run 以更新文案\n' + output);
            }

            // 到这里可以主动停止
            throw stopError;
          },
        },
      };

      // 开始执行
      new AI18n(config).run().catch(e => {
        // 主动停止
        if (e === stopError) return;

        // 异常停止要 exit 1
        console.error(e);
        process.exit(1);
      });
    }
  )
  .command(
    'run',
    '执行 i18n',
    _yargs => {},
    argv => {
      const config = getUserConfig(argv.config as string);

      // 开始执行
      new AI18n(config).run().catch(e => {
        console.error(e);
        process.exit(1);
      });
    }
  )
  .option('verbose', {
    alias: 'v',
    default: false,
  })
  .option('config', {
    default: './ai18n.json',
  }).argv;
