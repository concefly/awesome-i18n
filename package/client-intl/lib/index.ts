import IntlMessageFormat from 'intl-messageformat';

export interface MessageType {
  [key: string]: string;
}

/** 
 * 创建多语言 api
 */
export const createIntl = (
  opt: {
    message: MessageType;
    lang: string;
  },
  assignTo?
) => {
  // 通用翻译 api
  const __ = (text: string, description?: string, values?: any) => {
    try {
      const msg = new IntlMessageFormat(opt.message[text] || text, opt.lang);
      return msg.format({
        description,
        ...values,
      });
    } catch (e) {
      return text;
    }
  };

  // 用于定义多语言文案的空函数
  const __define = (text: string, description?: string) => void 0;

  const api = { __, __define };

  if (assignTo) {
    Object.assign(assignTo, api);
  }

  return api;
};
