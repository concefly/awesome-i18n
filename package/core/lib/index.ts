import * as glob from 'glob';
import * as _ from 'lodash';
import * as fs from 'fs';
import * as path from 'path';
import { ConfigType, getDefaultConfig } from './config';
import { LoaderConstructor } from '../type/loader';
import { ReducerConstructor, ReduceResultType, ReduceSourceType } from '../type/reducer';
import { TranslatorFuncType } from '../type/translator';

export interface OptType {
  hook?: {
    beforeLoad?: (filePath: string) => void;
    afterLoad?: (filePath: string, result: any) => void;
    afterLoadAll?: (result: any) => void;
    afterTranslate?: (src: string, result: string, from: string, to: string) => void;
  };
}

export default class AwesomeI18n {
  config: ConfigType = getDefaultConfig();
  opt: OptType;

  constructor(config: Partial<ConfigType> = {}, opt: OptType = {}) {
    // 合并自定义 config
    this.config = {
      ...this.config,
      ...config,
    };
    this.opt = opt;
  }

  getFs() {
    return fs;
  }

  getInputFiles() {
    return glob.sync(this.config.input);
  }

  readFile(filePath) {
    return this.getFs().readFileSync(filePath, 'utf-8');
  }

  getDumpFilePath(lang: string) {
    if (!this.getFs().existsSync(this.config.output)) {
      this.getFs().mkdirSync(this.config.output);
    }
    const filePath = path.join(this.config.output, `${lang}.json`);
    if (!this.getFs().existsSync(filePath)) {
      this.getFs().writeFileSync(filePath, '{}', 'utf-8');
    }
    return filePath;
  }

  getTranslator(): TranslatorFuncType {
    const { translator } = this.config;
    if (typeof translator === 'string') {
      return require(require.resolve(translator));
    }

    if (typeof translator === 'function') {
      return translator;
    }

    if (Array.isArray(translator)) {
      const func = require(require.resolve(translator[0]));
      return (...props) => func(...props, translator[1]);
    }

    throw new Error(`translator 定义有误`);
  }

  /**
   * 翻译到指定语言
   */
  async translate(
    reduceResult: ReduceResultType,
    lang: string
  ): Promise<{ [key: string]: string }> {
    const { defaultLang } = this.config;
    const translator = this.getTranslator();
    // 所有需要翻译的 keyword
    const keywords = _.chain(reduceResult)
      .map(r => r.keywords)
      .flatten()
      .uniq()
      .value();
    // 按语言分组串行翻译
    const chunkList = _.chunk(keywords, 5);
    const keywordMap: { [key: string]: string } = {};
    for (const chunk of chunkList) {
      const temp = await Promise.all(
        _.map(chunk, k => {
          // 若目标语言 === 默认语言，则原样返回
          if (defaultLang === lang)
            return {
              message: k,
              origin: k,
            };
          return translator(k, { from: defaultLang, to: lang })
            .then(r => ({
              ...r,
              origin: k,
            }))
            .catch(e => {
              console.error(e);
              return {
                message: k,
                origin: k,
              };
            });
        })
      );
      _.forEach(temp, t => {
        keywordMap[t.origin] = t.message;
        this.opt.hook &&
          this.opt.hook.afterTranslate &&
          this.opt.hook.afterTranslate(t.origin, t.message, defaultLang, lang);
      });
    }
    // 替换翻译文案
    const translatedResult = _.mapValues(reduceResult, r => {
      let message = r.message;
      _.forEach(r.keywords, k => {
        message = message.replace(k, keywordMap[k]);
      });
      return message;
    });
    return translatedResult;
  }

  /**
   * 导出多语言文件
   */
  dump(result: { [key: string]: string }, lang: string) {
    this.getFs().writeFileSync(this.getDumpFilePath(lang), JSON.stringify(result, null, 2), {
      encoding: 'utf-8',
    });
  }

  async run() {
    const inputFiles = this.getInputFiles();

    // load，提取所有多语言文案
    const loaderResultList = _.chain(inputFiles)
      .map(filePath => {
        const loader = _.find(this.config.loader, ({ test }) => test.test(filePath));
        if (!loader) throw new Error(`${filePath} 找不到对应的 loader`);

        const Loader: LoaderConstructor = require(require.resolve(loader.use)).default;

        this.opt.hook && this.opt.hook.beforeLoad && this.opt.hook.beforeLoad(filePath);

        const parseResult = new Loader().parse(this.readFile(filePath), filePath);

        this.opt.hook && this.opt.hook.afterLoad && this.opt.hook.afterLoad(filePath, parseResult);

        return parseResult;
      })
      .flatten()
      .value();

    this.opt.hook && this.opt.hook.afterLoadAll && this.opt.hook.afterLoadAll(loaderResultList);

    // reduce，合并多语言文案
    const Reducer: ReducerConstructor = require(require.resolve(this.config.reducer)).default;
    const reducerIns = new Reducer();

    let finalMap: {
      [lang: string]: {
        [key: string]: string;
      };
    } = {};

    // 循环处理每种语言
    for (const lang of this.config.langs) {
      const srcResultList = reducerIns.extract(
        JSON.parse(this.readFile(this.getDumpFilePath(lang)))
      );
      const reduceResult = reducerIns.reduce(_.map(loaderResultList, r => r.mark), srcResultList);

      const translateResult = await this.translate(reduceResult, lang);

      // 字段排序
      const sortedTranslateResult = _.pick(
        translateResult,
        _.chain(translateResult)
          .keys()
          .sort()
          .value()
      );

      this.dump(sortedTranslateResult, lang);
      finalMap[lang] = sortedTranslateResult;
    }

    return finalMap;
  }
}
