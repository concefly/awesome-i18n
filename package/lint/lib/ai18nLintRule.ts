import * as ts from 'typescript';
import * as Lint from 'tslint';
import { cnReg } from 'ai18n-type';

export interface IOptions {
  ignores?: string[];
  forces?: string[];
}

export function walk(ctx: Lint.WalkContext<void>) {
  // 跳过非 tsx 文件
  if (!ctx.sourceFile.fileName.endsWith('tsx')) return;

  // 检查 jsx 里的中文字面量
  const visitStringLiteral = (n: ts.StringLiteral) => {
    let text = n.text;
    if (!text.match(cnReg)) return;

    const parent = n.parent;
    if (
      ts.isCallExpression(parent) &&
      ['__', '__define'].indexOf(parent.expression.getText()) >= 0
    ) {
      // 跳过 __ 和 __define
      return;
    }

    text = text.trim();

    if (ts.isJsxAttribute(n.parent)) {
      // jsx attr 里的，形如 <div id="西瓜" />
      const fix = new Lint.Replacement(n.getStart(), n.getWidth(), `{__('${text}')}`);
      ctx.addFailureAtNode(n, `"${text}" 未做国际化处理`, fix);
      return;
    }

    const fix = new Lint.Replacement(n.getStart(), n.getWidth(), `__('${text}')`);
    ctx.addFailureAtNode(n, `"${text}" 未做国际化处理`, fix);
  };

  // 检查 <div>中文</div> 这种
  const visitJsxText = (n: ts.JsxText) => {
    const text = n.getText();
    if (!text || !cnReg.test(text)) return;

    const fix = new Lint.Replacement(n.getStart(), n.getWidth(), `{ __('${text.trim()}') }`);
    ctx.addFailureAtNode(n, `"${text.trim()}" 未做国际化处理`, fix);
  };

  const visitNode = (n: ts.Node) => {
    if (ts.isJsxText(n)) visitJsxText(n);
    if (ts.isStringLiteral(n)) visitStringLiteral(n);
    else walkChildren(n);
  };

  const walkChildren = (n: ts.Node) => ts.forEachChild(n, visitNode);
  walkChildren(ctx.sourceFile);
}

export class Rule extends Lint.Rules.AbstractRule {
  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
    const options = this.getOptions();
    const { ignores = [], forces = [] } = options.ruleArguments[0] as IOptions;

    const isInIgnoreList = ignores.some(k => sourceFile.fileName.includes(k));
    const isInForceList = forces.some(k => sourceFile.fileName.includes(k));

    const shouldIgnore = isInIgnoreList && !isInForceList;
    if (shouldIgnore) return [];

    return this.applyWithFunction(sourceFile, walk);
  }
}
