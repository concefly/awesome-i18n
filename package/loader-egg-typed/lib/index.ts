import { walk } from './walk';
import * as ts from 'typescript';

export interface LoaderConfigType {}
export interface LoaderParseType {
  mark: {
    key: string;
    description?: string;
  };
}

export const METHOD_LIST = [
  'BadRequestError',
  'ForbiddenError',
  'ServerError',
  'NotFoundError',
  'UnauthorizedError',
];

class Loader {
  public config: LoaderConfigType;
  public ast: ts.SourceFile;

  constructor(config: LoaderConfigType = {}) {
    this.config = { plugins: { jsx: true }, ...config };
  }

  private parseCallExpression(node: ts.CallExpression): LoaderParseType {
    const method = node.expression.getText();
    if (METHOD_LIST.indexOf(method) >= 0) {
      const [arg1] = node.arguments;

      // 形如： BadRequestError('非法请求')
      if (arg1 && ts.isStringLiteral(arg1)) {
        return {
          mark: {
            key: arg1.getText().replace(/^"|^'|"$|'$/g, ''),
          },
        };
      }

      // 形如： BadRequestError({ code: -1, message: '需要人机验证', type: 'afs' });
      if (arg1 && ts.isObjectLiteralExpression(arg1)) {
        const messageProperty = arg1.properties.find(
          p =>
            ts.isPropertyAssignment(p) &&
            ts.isIdentifier(p.name) &&
            p.name.getText() === 'message' &&
            ts.isStringLiteral(p.initializer)
        );
        if (messageProperty && ts.isPropertyAssignment(messageProperty)) {
          return {
            mark: {
              key: messageProperty.initializer.getText().replace(/^"|^'|"$|'$/g, ''),
            },
          };
        }
      }
    }
  }

  parse(code: string, filePath: string): LoaderParseType[] {
    const source = (this.ast = ts.createSourceFile(
      filePath,
      code,
      ts.ScriptTarget.ES2015,
      /*setParentNodes */ true
    ));

    const markList: LoaderParseType[] = [];

    walk(source, node => {
      if (ts.isCallExpression(node)) {
        const r = this.parseCallExpression(node);
        if (r) markList.push(r);
      }
    });

    return markList;
  }
}

export default Loader;
