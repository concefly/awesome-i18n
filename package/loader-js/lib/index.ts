import walk from './walk';
import { Parser as OriginParser, Options } from 'acorn';
import * as acornJsx from 'acorn-jsx';
import { Program, CallExpression, Literal, Identifier } from 'estree';

const Parser = OriginParser.extend(acornJsx(OriginParser));

export interface LoaderConfigType {}
export interface LoaderParseType {
  mark: {
    key: string;
    description?: string;
  };
}

class Loader {
  static PARSER_CONFIG: Options = {
    sourceType: 'module',
    ecmaVersion: 2019,
  };

  public config: LoaderConfigType;
  public ast: Program;

  constructor(config: LoaderConfigType = {}) {
    this.config = { plugins: { jsx: true }, ...config };
  }

  private parseCallExpression(node: CallExpression) {
    const calleeName = (node.callee as Identifier).name;
    // 提取 __() 和 __define() 定义的文案
    if (['__', '__define'].indexOf(calleeName) >= 0) {
      const args = node.arguments || [];
      if (args.length === 0) return;
      // 参数 0 必须是纯文本
      if (
        args[0].type === 'Literal' &&
        (args[0] as Literal).value &&
        typeof (args[0] as Literal).value === 'string'
      ) {
        const mark: LoaderParseType['mark'] = {
          key: (args[0] as Literal).value + '',
        };
        if (
          args[1] &&
          args[1].type === 'Literal' &&
          (args[1] as Literal).value &&
          typeof (args[1] as Literal).value === 'string'
        ) {
          mark.description = (args[1] as Literal).value as string;
        }
        return { mark };
      }
    }
  }

  parse(code: string): LoaderParseType[] {
    const markList: LoaderParseType[] = [];
    const ast = (this.ast = new Parser(Loader.PARSER_CONFIG, code).parse() as any);
    walk.simple(ast, {
      CallExpression: node => {
        const r = this.parseCallExpression(node);
        if (r) markList.push(r);
      },
    });
    return markList;
  }
}

export default Loader;
