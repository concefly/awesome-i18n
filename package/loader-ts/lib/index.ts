import { walk } from './walk';
import * as ts from 'typescript';

export interface LoaderConfigType {}
export interface LoaderParseType {
  mark: {
    key: string;
    description?: string;
  };
}

class Loader {
  public config: LoaderConfigType;
  public ast: ts.SourceFile;

  constructor(config: LoaderConfigType = {}) {
    this.config = { plugins: { jsx: true }, ...config };
  }

  private parseCallExpression(node: ts.CallExpression): LoaderParseType {
    const method = node.expression.getText();
    if (['__', '__define'].indexOf(method) >= 0) {
      const [key, description] = node.arguments;
      // 只提取 key 为字符串字面量的 token
      const normalizedKey =
        key && ts.isStringLiteral(key) && key.getText().replace(/^"|^'|"$|'$/g, '');
      const normalizedDesc =
        description &&
        ts.isStringLiteral(description) &&
        description.getText().replace(/^"|^'|"$|'$/g, '');
      if (normalizedKey) {
        const re: LoaderParseType = {
          mark: {
            key: normalizedKey,
          },
        };
        if (normalizedDesc) {
          re.mark.description = normalizedDesc;
        }
        return re;
      }
    }
  }

  parse(code: string, filePath: string): LoaderParseType[] {
    const source = (this.ast = ts.createSourceFile(
      filePath,
      code,
      ts.ScriptTarget.ES2015,
      /*setParentNodes */ true
    ));

    const markList: LoaderParseType[] = [];

    walk(source, node => {
      if (ts.isCallExpression(node)) {
        const r = this.parseCallExpression(node);
        if (r) markList.push(r);
      }
    });

    return markList;
  }
}

export default Loader;
