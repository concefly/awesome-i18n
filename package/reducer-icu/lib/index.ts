import * as _ from 'lodash';
import { parse, TokenType, MessageTextElement, ArgumentElement, build } from './icu';
import { buildSimpleSelectPattern, buildSimpleTextPattern, mergeValues } from './util';

export interface MapType<T> {
  [key: string]: T;
}
export interface ConfigType {}
export interface SourceType {
  key: string;
  description?: string;
  message?: string;
}
export interface ResultType {
  [key: string]: {
    message: string;
    /** 待翻译的关键字列表 */
    keywords: string[];
  };
}

// const EMPTY_DESCRIPTION = Symbol('none');
const EMPTY_DESCRIPTION = 'other';

export default class Reducer {
  config: ConfigType;

  constructor(config: ConfigType = {}) {
    this.config = config;
  }

  makeStructure(src: SourceType[]) {
    return (
      _.chain(src)
        .groupBy('key')
        // 补全 description，方便后续 groupBy
        .mapValues(list =>
          _.map(list, item => {
            const re = { ...item };
            if (!re.description) re.description = EMPTY_DESCRIPTION;
            return re;
          })
        )
        .mapValues((payloadList: SourceType[]) =>
          _.chain(payloadList)
            .groupBy('description')
            .mapValues(_.first)
            .value()
        )
        .value()
    );
  }

  extract(raw: { [key: string]: string }): SourceType[] {
    return _.chain(raw)
      .mapValues((value: string, key: string) => {
        const ast = parse(value);
        if (
          ast.elements.length > 1 &&
          ast.elements.some(
            e => e.type === TokenType.argumentElement && !!(e as ArgumentElement).format
          )
        )
          throw new Error('参数选择器存在时，只能有一个 element');
        // 上下文动态文本模式
        if (
          ast.elements[0].type === TokenType.argumentElement &&
          (ast.elements[0] as ArgumentElement).format.type === TokenType.selectFormat
        ) {
          return (ast.elements[0] as ArgumentElement).format.options.map(opt => {
            return {
              key,
              description: opt.selector,
              message: build(opt.value),
            };
          });
        }
        // 原样返回兜底
        return [
          {
            key,
            message: value,
          },
        ];
      })
      .mapValues(v => _.flatten(v))
      .reduce((r, v) => r.concat(v), [])
      .value();
  }

  reduce(incoming: SourceType[], src: SourceType[]): ResultType {
    const incomingStructure = this.makeStructure(incoming);
    const srcStructure = this.makeStructure(src);
    // 两重 merge
    // 1. 合并 key
    const resolvedStructure = mergeValues<MapType<SourceType>>(
      srcStructure,
      incomingStructure,
      (srcDescMap, incomingDescMap) =>
        // 2. 合并上下文
        mergeValues<SourceType>(srcDescMap, incomingDescMap, (s, i) => {
          // source 有 message，则原样返回(不修改源文件的翻译)
          if (s.message) return s;
          else return i;
        })
    );
    // 合并 descMap
    const re = _.mapValues(resolvedStructure, (descMap, key) => {
      const descList = _.keys(descMap);
      // 只有一个 other 上下文，则 message = incomingMessage || key
      if (descList.length === 1 && descList[0] === EMPTY_DESCRIPTION) {
        const incomingMessage = descMap[EMPTY_DESCRIPTION].message;
        const keywords = incomingMessage ? [] : [key];
        const message = incomingMessage || key;
        return { keywords, message };
      } else {
        const keywords: string[] = [];
        const ast = buildSimpleSelectPattern(
          'description',
          _.map(descMap, (s, desc) => {
            // source 里没有 message，则要把 incoming 的 key 加入翻译列表
            if (!s.message) keywords.push(key);
            return {
              selector: desc,
              message: s.message || key,
            };
          })
        );
        return {
          keywords,
          message: build(ast),
        };
      }
    });
    return re;
  }
}
