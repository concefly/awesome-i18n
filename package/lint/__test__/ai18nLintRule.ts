import { WalkContext } from 'tslint';
import { walk } from '../lib/ai18nLintRule';
import * as ts from 'typescript';

const createLintCtx = (content: string) => {
  const options: ts.CompilerOptions = { jsx: ts.JsxEmit.None };
  const host = ts.createCompilerHost(options, true);
  host.readFile = () => content;

  const p = ts.createProgram(['mock.tsx'], options, host);
  const sourceFile = p.getSourceFile('mock.tsx')!;

  return new WalkContext<any>(sourceFile, 'x', undefined);
};

const inspectCtx = (ctx: WalkContext) => {
  const result = ctx.failures.map(f => {
    return [f.getFailure(), f.getFix()];
  });

  return result;
};

describe('ai18nLintRule', () => {
  it('jsx element', () => {
    const ctx = createLintCtx('const a = <div title="标题">苹果<span title="标签">{"香蕉"}</span></div>');
    walk(ctx);
    expect(inspectCtx(ctx)).toMatchSnapshot();
  });

  it('jsx StringLiteral', () => {
    const ctx = createLintCtx('const a = "苹果"');
    walk(ctx);
    expect(inspectCtx(ctx)).toMatchSnapshot();
  });

  it('跳过非全中文字符串', () => {
    const ctx = createLintCtx('const a = "苹果 xxx"');
    walk(ctx);
    expect(ctx.failures.length === 0);
  });

  it('跳过已被 __ 和 __define 包裹的', () => {
    const ctx = createLintCtx(`const a = __("苹果") + __define("香蕉") + foo("西瓜")`);
    walk(ctx);
    expect(inspectCtx(ctx)).toMatchSnapshot();
  });
});
