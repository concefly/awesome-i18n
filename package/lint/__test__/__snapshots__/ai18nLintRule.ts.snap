// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ai18nLintRule jsx StringLiteral 1`] = `
Array [
  Array [
    "\\"苹果\\" 未做国际化处理",
    Replacement {
      "length": 4,
      "start": 10,
      "text": "__('苹果')",
    },
  ],
]
`;

exports[`ai18nLintRule jsx element 1`] = `
Array [
  Array [
    "\\"标题\\" 未做国际化处理",
    Replacement {
      "length": 4,
      "start": 21,
      "text": "{__('标题')}",
    },
  ],
  Array [
    "\\"苹果\\" 未做国际化处理",
    Replacement {
      "length": 2,
      "start": 26,
      "text": "{ __('苹果') }",
    },
  ],
  Array [
    "\\"标签\\" 未做国际化处理",
    Replacement {
      "length": 4,
      "start": 40,
      "text": "{__('标签')}",
    },
  ],
  Array [
    "\\"香蕉\\" 未做国际化处理",
    Replacement {
      "length": 4,
      "start": 46,
      "text": "__('香蕉')",
    },
  ],
]
`;

exports[`ai18nLintRule 识别中文符号 1`] = `
Array [
  Array [
    "\\"苹果。，、？！；“”《》（）\\" 未做国际化处理",
    Replacement {
      "length": 16,
      "start": 10,
      "text": "__('苹果。，、？！；“”《》（）')",
    },
  ],
]
`;

exports[`ai18nLintRule 跳过已被 __ 和 __define 包裹的 1`] = `
Array [
  Array [
    "\\"西瓜\\" 未做国际化处理",
    Replacement {
      "length": 4,
      "start": 42,
      "text": "__('西瓜')",
    },
  ],
]
`;

exports[`ai18nLintRule 跳过非全中文字符串 1`] = `Array []`;
